runtime/domain.c

```diff
@@ -1106,6 +1108,35 @@ int caml_domain_is_in_stw(void) {
 }
 #endif
 
+/* During a stop-the-world (STW), all currently running domains stop
+   their usual work and synchronize to all call the same function.
+
+   STW sections use [all_domains_lock] and a condition variable
+   [all_domains_cond] to guarantee that no domain is running something
+   else:
+
+   - If two STW sections are attempted in parallel, only one will
+     manage to take the lock, and the domain starting the other will
+     join that winning STW section, without running its own STW code
+     at all. (This is the [_try_] in the function name: if it returns
+     0, the STW section did not run at all, so you should call this
+     function in a loop.)
+
+   - Domain creation code will not run in parallel with a STW section:
+     it waits for the [all_domains_cond] that is signalled at the end
+     of the STW section (see [create_domain]).
+
+   - Domain termination code will not run in parallel with a STW
```

Comment:
Not sure I would write it this way; code in `domain_terminate` will run. In fact
a terminating domain must take part in STW sections until it is in a state where it 
can safely leave the STW set. 

My focus would be:
```
A terminating domain uses [all_domains_lock] in [domain_terminate] to ensure 
it does not drop itself from the STW set while a STW section is in progress. 
```

Reply:
What I had in mind with "domain termination code" is the part that actually cleanups the domain state and releases resources. Your proposal is of course correct, but it's also important to note that there cannot be a race between the STW code and the domain cleanup (because the cleanup happens with `all_domains_lock` hold, so new STW sections cannot start). I will try to rephase.

Generally my goal was to understand what it means for some mutable variables to be "protected by STW sections". STW sections act as a mutual-exclusion mechanism, what sort of synchronization guarantees does it provide? (I had to think about this in the context of #10971.)

Reply:
True, domain cleanup is done once the domain is out of the STW set. 

At the top of `domain.c`, I had a go at listing the STW invariants:
https://github.com/ocaml/ocaml/blob/04ddddd0e1d2bf2acb5091f434b93387243d4624/runtime/domain.c#L48-L62

Reply:
Good point! I need an additional invariant, which is that domain initialization and cleanup never run concurrently with STW sections (in addition to mutator code). This makes sure that STW sections can change domain parameters without racing against domain_create or terminate_domain.

Question:

> they are within a section no mutator code is running 

What does this one mean? (Maybe "a section *where* no mutator code is running"?)

Reply:
Agree, could be clearer, maybe:
```
- all domains know that there is no mutator code running in the section
- domains take care to not trigger callbacks (e.g. a finaliser or signal 
handler) from within the section
```

Reply:
I pushed a new commit that rephrases this toplevel explanation, would you care to give your opinion? This is just a proposal for clarity (and just a small bit of extra information), it's not about correctness, so I'm perfectly fine with leaving things as-is (or taking your proposal exactly), just tell me.